source/material_model/melt_simple.cc

```diff
@@ -258,97 +236,110 @@ namespace aspect
-                  melting_rate = eq_melt_fraction - old_porosity[i];
+                  porosity_change = eq_melt_fraction - old_porosity;
                 }
               else
-                // batch melting
-                melting_rate = melt_fraction(in.temperature[i], this->get_adiabatic_conditions().pressure(in.position[i]))
-                               - std::max(maximum_melt_fractions[i], 0.0);
+                {
+                  // batch melting
+                  porosity_change = melt_fraction(in.temperature[i], this->get_adiabatic_conditions().pressure(in.position[i]))
+                                    - std::max(maximum_melt_fraction, 0.0);
+                  porosity_change = std::max(porosity_change, 0.0);
+
+                  // freezing of melt below the solidus
+                  {
+                    // If the porosity is larger than the equilibrium melt fraction, melt should freeze again.
+                    // Because we do not track the melt composition, we have to use a workaround here for freezing of melt:
+                    // We reduce the porosity until either it reaches the equilibrium melt fraction, or the depletion
+                    // (peridotite field), which decreases as melt freezes, reaches the same value as the equilibrium
+                    // melt fraction, whatever happens earlier. An exception is when the melt fraction is zero; in this case
+                    // all melt should freeze.
+                    const double eq_melt_fraction = melt_fraction(in.temperature[i], this->get_adiabatic_conditions().pressure(in.position[i]));
+
+                    // If the porosity change is not negative, there is no freezing, and the change in porosity
+                    // is covered by the melting relation above.
+
+                    // porosity reaches the equilibrium melt fraction:
+                    const double porosity_change_wrt_melt_fraction = std::min(eq_melt_fraction - old_porosity - porosity_change,0.0);
+
+                    // depletion reaches the equilibrium melt fraction:
+                    const double porosity_change_wrt_depletion = std::min((eq_melt_fraction - std::max(maximum_melt_fraction, 0.0))
+                                                                          * (1.0 - old_porosity) / (1.0 - maximum_melt_fraction),0.0);
+                    double freezing_amount = std::max(porosity_change_wrt_melt_fraction, porosity_change_wrt_depletion);
+
+                    if (eq_melt_fraction == 0.0)
+                      freezing_amount = - old_porosity;
+
+                    porosity_change += freezing_amount;
+
+                    // Adapt time scale of freezing with respect to melting.
+                    if (porosity_change < 0 )
+                      porosity_change *= freezing_rate * melting_time_scale;
```

Comment:
Can you add a comment here why you multiply by both? "E.g. we divide by the melting_time_scale later on in any case, but this is not necessary for this case, thus ...". That would make it easier to understand.